workflows/release-binaries: Enable Windows x86 builds

tstellar opened the pull request:
This is meant to be a drop-in replacement for binaries built with the build_llvm_release.bat script, though the cmake configuration is slightly different.  These builds also have flang and mlir which the script does not and it also enables all targets.

tstellar committed:
workflows/release-binaries: Enable Windows x86 builds

tstellar committed:
Fix typo

tstellar committed:
Disable unsupported runtimes

whatisthis-dot commented:
💌

whatisthis-dot commented:
> 💌

tstellar committed:
Windows-2022

tstellar committed:
Windows fixes

tstellar committed:
XXX: Debug

tstellar committed:
Fix

tstellar committed:
Update Release.cmake

tstellar committed:
Update release-binaries.yml

tstellar committed:
Update release-binaries.yml

tstellar committed:
Update release-binaries.yml

tstellar committed:
Use powershell for windows

tstellar committed:
Fix

tstellar committed:
Fix

tstellar committed:
Fix

tstellar committed:
XXX:Debug

tstellar committed:
XXX: Debug

tstellar committed:
Fix

tstellar committed:
Fix

tstellar committed:
Check diskspace

tstellar committed:
Debug

tstellar committed:
Debug

tstellar committed:
Fix

tstellar committed:
Debug

tstellar committed:
Debug

tstellar committed:
Debug

tstellar committed:
Debug

tstellar committed:
Debug

tstellar committed:
bash fix

tstellar committed:
Fix space n cmake

tstellar committed:
Fix

tstellar committed:
Fix

tstellar committed:
Fix

tstellar committed:
Fix

tstellar committed:
Fix

tstellar committed:
Merge branch 'main' into release-windows

tstellar committed:
Try to use response file

tstellar committed:
Fix path lengths

tstellar committed:
try to fix long paths

tstellar committed:
Fix typo

tstellar committed:
Use sbust to shorten paths

tstellar committed:
Fix path shortening

tstellar committed:
Fix typo

tstellar committed:
Fix subst

tstellar committed:
Fix

tstellar committed:
Fix

tstellar committed:
Debugging

tstellar committed:
Use powershell for everything

tstellar committed:
Use backticks

tstellar committed:
Debugging fixes

tstellar committed:
Try to fix installer size on Windows

tstellar committed:
Switch to wix and some debugging

tstellar committed:
FIxes and ebugging

tstellar committed:
Wix does not like strings on versions

tstellar committed:
Remove nsis log

tstellar committed:
Remove nsis log

tstellar committed:
Fix suffix

tstellar committed:
Drop log

tstellar committed:
Mv artifacts

tstellar committed:
Merge remote-tracking branch 'origin/main' into release-windows

tstellar committed:
Cleanups last commit was working

tstellar committed:
More cleanups previous build working

tstellar committed:
Simplify more previous build works

tstellar committed:
workflows/release-binaries: Run tests on the same runner as the build

Also, ignore the test results since they almost always fail.  This
allows us to simplify the build process and skip uploading and
downloading the build and source directories which are huge.

tstellar committed:
Add missing shell parameter

tstellar committed:
Remove some dead cone

tstellar committed:
Add back in llvm checkout

tstellar committed:
Merge remote-tracking branch 'tstellar/release-binaries-cleanup' into release-windows

tstellar committed:
Fix shell for tests

tstellar committed:
Remove some debug statements from lit

tstellar committed:
Merge remote-tracking branch 'origin/main' into release-windows

tstellar committed:
MOre fixes previous is working

tstellar committed:
Enable more projects

tstellar committed:
Fix cmake command

tstellar committed:
Build all projects

tstellar committed:
XXX: Enable lto and use correct runners

tstellar committed:
Dump logs

tstellar committed:
Check diskspace

tstellar committed:
Fix typo

tstellar committed:
Disable LTO

tstellar committed:
Build c dylib on windows

tstellar committed:
Merge remote-tracking branch 'origin/main' into release-windows

tstellar committed:
Cleanups

boomanaiden154 reviewed (approved):
GHA side LGTM.

Is the plan to deprecate the `build_llvm_release.bat` script once we're using this workflow to produce our Windows binaries too?

I'm assuming at some point we want to look at doing Windows on ARM builds too now that there are runners available for that?

zmodem reviewed (commented):
On the one hand I'm excited about this, because long-term automatically built release binaries are obviously the right thing.

On the other hand, I'm concerned because this is changing a lot of variables all at once.

> This is meant to be a drop-in replacement for binaries built with the build_llvm_release.bat script

That objective doesn't seem compatible with all the changes:

- Disabling LTO -- this seems bad for performance?
- Does the new process include PGO, and how is training done?
- Switch from the old "PROJECTS=...;compiler-rt" to the "runtimes build"
- Building a Wix .msi instead of NSIS .exe, which is also extractable by 7zip. Some people rely on this, though I think moving to shipping both an installer and .zip file is the sweet spot.
- "Run tests" say "These almost always fail, don't block on them", whereas the old script blocks on (most) failures. We don't want to ship binaries with failing tests.

This also moves the release logic into .yaml. An important property of the .bat file is that we can point people to it when they ask how a release is built, and they can reproduce it.

I think the ideal way to do this would be to:

- Gradually and carefully make the release packaging changes we want (include more targets, include flang, switch to Wix, use Release.cmake, ...) to the .bat file
- Have the Github workflow run the .bat file, so users can easily reproduce the release packages.

tstellar committed:
Stop using wix

tstellar committed:
Debug

tstellar committed:
Reduce targets to fix build with nsis

tstellar committed:
toolchain only

tstellar committed:
Strip moreout

tstellar committed:
Reduce to test nsis

tstellar committed:
Switch back to exe

tstellar committed:
Match config

tstellar committed:
[CMake][Release] Stop linking against stage1 runtimes

This was causing a build failure on Darwin and a test failure
on Linux.  This config is not widely used or well tested, so
I don't think the potential and likely small performance gains
and the portability from this are worth the maintenance costs.

tstellar committed:
Change variable name

tstellar committed:
Merge remote-tracking branch 'tstellar/release-use-system-libs' into release-windows

tstellar committed:
Install libxml2

tstellar committed:
fix typo

tstellar committed:
Fix typo

tstellar committed:
Don't use bash for libxml2

tstellar committed:
Fix line breaks

tstellar committed:
Fix line endigns

tstellar committed:
Fix powershellstuff

tstellar committed:
Remove whitespace

tstellar committed:
libxml fix

tstellar committed:
Fix xml paths

tstellar committed:
fix paths

tstellar committed:
Add debugging

tstellar committed:
Fix

tstellar committed:
debug

tstellar committed:
Try to fix libxml detection

tstellar committed:
Fixes

tstellar committed:
Debug

tstellar committed:
Fix upload

tstellar committed:
xxx; debug

tstellar committed:
More debug

tstellar committed:
Static link

tstellar committed:
Merge remote-tracking branch 'origin/main' into HEAD

tstellar committed:
Debug/fix

tstellar committed:
Simplify libxml2

tstellar committed:
disable premerge

tstellar committed:
Fix typo

tstellar committed:
libxml2 with package manager

tstellar committed:
Fix

tstellar committed:
Fix cmake loookpu

tstellar committed:
More debug

tstellar committed:
Fixes

tstellar committed:
libxml changes

tstellar committed:
Fix typo

tstellar committed:
XXX:

tstellar committed:
Fix typo

tstellar committed:
Fix arguments

tstellar committed:
Add includes

tstellar committed:
Debug

tstellar committed:
Switch to custom built

tstellar committed:
Fixes

tstellar committed:
Attemptto fix

tstellar committed:
Attemptto fix

tstellar committed:
Fix quotes

tstellar committed:
Remove more quotes

tstellar committed:
more fixes

tstellar committed:
Fix paths

tstellar committed:
Install libxml2

tstellar committed:
Debug

tstellar committed:
Add debug

tstellar committed:
Fix upload

tstellar committed:
Install static library

tstellar committed:
Run on github

tstellar committed:
Build relase only

tstellar committed:
Fix static install

tstellar committed:
Fix library name

tstellar committed:
Xheck library

tstellar committed:
Try to build static libs

tstellar committed:
Remove debug

tstellar committed:
FIx static

tstellar committed:
Fix typo

tstellar committed:
Use self-built libxml2

tstellar committed:
Fix library name

tstellar committed:
Try setting Definitions

tstellar committed:
Remove extra upload

tstellar committed:
Try unsetting Definitions

tstellar committed:
Rfactor

tstellar committed:
cleanups

tstellar committed:
depot

tstellar committed:
Fix typo

tstellar committed:
Debug vcpkg

tstellar committed:
Debug

tstellar committed:
Debug

tstellar committed:
Debug

tstellar committed:
Debug

tstellar committed:
Attempt to build with vcpkg

tstellar committed:
Fix typo

tstellar committed:
Merge remote-tracking branch 'origin/main' into HEAD

tstellar committed:
Full fix

tstellar committed:
Try to install static vcpkg

tstellar committed:
Fix typo

tstellar committed:
Try to update

tstellar committed:
Fix typo

tstellar committed:
Try to force static libs

tstellar committed:
More debug

tstellar committed:
More debug

tstellar committed:
Last try for static vcpkg

tstellar committed:
Fix typo

tstellar committed:
Fixes

tstellar committed:
Fix library name

tstellar committed:
Fix build

tstellar committed:
Attempt again

tstellar committed:
Don"t force libraries

tstellar committed:
Disable features

tstellar committed:
Attempt automatic detection

tstellar committed:
Debug

tstellar committed:
Fix typo

tstellar committed:
Debug

tstellar committed:
Fix debug

tstellar committed:
Fix

tstellar committed:
Debug

tstellar committed:
Debug

tstellar committed:
Debug

tstellar committed:
Change prefix path

tstellar committed:
XXX: Config mode

tstellar committed:
Remove some debugging output

tstellar committed:
Use LibXml2Dir

tstellar committed:
Use vpckg

tstellar committed:
Fixes

tstellar committed:
Test without the config option

tstellar committed:
Fixes

tstellar committed:
Disable flang

tstellar committed:
Drop targets

tstellar committed:
Switch to WIX

tstellar committed:
Add wix log

tstellar committed:
Fix wix version

tstellar committed:
Fix

tstellar committed:
Disable pipefail

tstellar committed:
more debug

tstellar committed:
Fix suffix

tstellar committed:
Use msi

tstellar committed:
disable autocrlf

tstellar committed:
Some fixes

tstellar committed:
Merge remote-tracking branch 'origin/main' into HEAD

tstellar committed:
Setu python

tstellar committed:
Fix python

tstellar committed:
Add full commit

tstellar committed:
Use specific python interpreter for pip

tstellar committed:
Fix path

tstellar committed:
Disable lldb tests

tstellar committed:
Cleanups

tstellar committed:
Fix projects and runtimes

zmodem commented:
> Logs here:
> https://github.com/llvm/llvm-project/actions/runs/19373045853/job/55433478607?pr=150793

The first error is:

```
2025-11-14T18:08:50.4596533Z FAILED: [code=1] openmp/runtime/src/CMakeFiles/omp.dir/kmp_csupport.cpp.obj 
2025-11-14T18:08:50.4601282Z S:\llvm\utils\release\llvm_package_21.1.4\build_amd64_stage0\bin\clang-cl.exe --target=x86_64-pc-windows-msvc  /nologo -TP -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_GLIBCXX_NO_ASSERTIONS -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Domp_EXPORTS -IS:\llvm\utils\release\llvm_package_21.1.4\build_amd64_stage0\runtimes\runtimes-bins\openmp\runtime\src -IS:\openmp\runtime\src -IS:\openmp\runtime\src\i18n -IS:\openmp\runtime\src\include -IS:\openmp\runtime\src\thirdparty\ittnotify /DWIN32 /D_WINDOWS   /Zc:inline /Zc:__cplusplus /Oi /Brepro /bigobj /permissive- -Werror=unguarded-availability-new /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd5105 -wd4324 -wd4251 -wd4275 -w14062 -we4238 /Gw -fcolor-diagnostics -Wcast-qual -Wformat-pedantic -Wimplicit-fallthrough -Wsign-compare -Wno-extra -Wno-pedantic -wd4201 -wd4190 /O2 /Ob2 /DNDEBUG -std:c++17 -MT   -D _CRT_SECURE_NO_WARNINGS -D _CRT_SECURE_NO_DEPRECATE -D _WINDOWS -D _WINNT -D _WIN32_WINNT=0x0501 -D _USRDLL -Wno-covered-switch-default -Wno-frame-address -Wno-strict-aliasing -Wno-switch -Wno-uninitialized -Wno-return-type-c-linkage -Wno-cast-qual -Wno-int-to-void-pointer-cast /GS /EHsc -mrtm /showIncludes /Foopenmp\runtime\src\CMakeFiles\omp.dir\kmp_csupport.cpp.obj /Fdopenmp\runtime\src\CMakeFiles\omp.dir\ -c -- S:\openmp\runtime\src\kmp_csupport.cpp
2025-11-14T18:08:50.4612059Z S:\openmp\runtime\src\kmp_csupport.cpp(1349,52): error: use of undeclared identifier 'omp_lock_hint_none'
2025-11-14T18:08:50.4612712Z  1349 |   __kmpc_critical_with_hint(loc, global_tid, crit, omp_lock_hint_none);
2025-11-14T18:08:50.4614233Z       |
```

I think `omp_lock_hint_none` is defined in `openmp/runtime/src/include/omp.h.var` which should get processed by cmake (openmp/runtime/src/CMakeLists.txt) into `omp.h` in clang's resource dir:

```
# The generated headers will be placed in clang's resource if using it to
# compile openmp in a runtimes-bootstrapping build.
if(LLVM_TREE_AVAILABLE AND LLVM_RUNTIMES_BUILD)
  set(LIBOMP_HEADERS_INTDIR ${LLVM_BINARY_DIR}/${LIBOMP_HEADERS_INSTALL_PATH})
else()
  set(LIBOMP_HEADERS_INTDIR ${CMAKE_CURRENT_BINARY_DIR})
endif()
```

I guess it somehow failed to generate the header, or at least didn't put it in the right place?

Is it possible to run with higher cmake verbosity to see if/where it writes that file?

Being paranoid: could there be some weird interaction between passing `--version 21.1.4` but building trunk which is 22.0.0? (imaybe openmp gets the version number in clang's resource dir wrong due to this somehow)

---

fwiw `llvm\utils\release\build_llvm_release.bat --version 22.0.0-rc0 --skip-checkout --x64 --force-msvc` seems to work for me on trunk (af45b0202cdd443beedb02392f653d8cff5bd931)

tstellar commented:
> Being paranoid: could there be some weird interaction between passing --version 21.1.4 but building trunk which is 22.0.0? (imaybe openmp gets the version number in clang's resource dir wrong due to this somehow)

I think this fixed the OpenMP error.  However, I'm starting to have problems with the python detection, but this seems like a change it's caused by a change in the GitHub environment recently and not with the script.
